Source/UI/IdentityUI.swift

```diff
@@ -300,27 +302,64 @@ public class IdentityUI {
             if let result = uiResult {
                 self?.delegate?.didFinish(result: result)
             }
         }
 
-        if self.navigationController.presentingViewController != nil {
-            // It might be that `IdentityUIViewController.endLoading()` has been called just before getting here, in case the result of a networking
-            // operation caused the flow to end. The `endLoading()` method will then trigger a `view.isUserInteractionEnabled = true`, which would cause the
-            // keyboard to show up again during the dismiss animation, resulting in a very weird and funky UI glitch. In order to avoid that, we force an
-            // `endEditing()` on the topmost view (if any) before starting the view dismissing.
-            self.navigationController.topViewController?.view.endEditing(true)
+        let dismissFlow = { [weak self] in
+
+            // This is no more the currently presented login flow
+            IdentityUI.presentedIdentityUI = nil
+
+            if self?.navigationController.presentingViewController != nil {
+                // It might be that `IdentityUIViewController.endLoading()` has been called just before getting here, in case the result of a networking
+                // operation caused the flow to end. The `endLoading()` method will then trigger a `view.isUserInteractionEnabled = true`, which would cause the
+                // keyboard to show up again during the dismiss animation, resulting in a very weird and funky UI glitch. In order to avoid that, we force an
+                // `endEditing()` on the topmost view (if any) before starting the view dismissing.
+                self?.navigationController.topViewController?.view.endEditing(true)
 
-            self.navigationController.dismiss(animated: true) {
+                self?.navigationController.dismiss(animated: true) {
+                    finish()
+                }
+            } else {
                 finish()
             }
+        }
+
+        let restartFlow = { [weak self, weak presentingViewController] in
+            guard let strongSelf = self else { return }
+            if strongSelf.navigationController.presentingViewController != nil {
+                strongSelf.navigationController.popToRootViewController(animated: true)
+            } else {
+                strongSelf.configuration.presentationHook?(strongSelf.navigationController)
+                presentingViewController?.present(strongSelf.navigationController, animated: true)
+            }
+        }
+
+        if let delegate = self.delegate, case let .completed(user)? = uiResult {
+            delegate.willSucceed(with: user, on: self.navigationController.topViewController) { disposition in
+                switch disposition {
+                case .continue:
+                    dismissFlow()
+                case let .failed(title: title, message: message):
+                    if self.navigationController.presentingViewController == nil {
```

Comment:
Why not present the navigationController and viewController before start doing anything? Or is there some part of the flow im missing.

Reply:
We might not need to if we come from a deep link to verify the auth code and the client says we are all good, so continue the flow.